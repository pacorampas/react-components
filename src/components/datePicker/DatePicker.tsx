import React, { useState } from 'react'
import { DateTime } from 'luxon'
import styled, { css } from 'styled-components'
import { MonthDays, DayArguments } from './components/MothDays'
import { VARIANT } from '../../theme'

export interface DatePickerProps {
  className?: string
  variant?: VARIANT
  selected?: Date
  fillEmptyDays?: boolean
  onChange?: ({ date }: { date: Date }) => void
}

const defaultProps = {
  variant: VARIANT.PRIMARY,
  selected: new Date(),
  fillEmptyDays: false,
}

const Wrapper = styled.div`
  width: 100%;
  min-width: 210px;
`

const Header = styled.header`
  display: flex;
  align-items: center;
`

const IconButton = styled.button`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 30px;
  height: 30px;
  padding: 0;
  margin: 0;
  background: none;
  border-color: transparent;
  border-radius: 3px;
  box-sizing: border-box;
  cursor: pointer;
  opacity: 0.7;
  transform: scale(0.9);

  ${() => css`
    &:hover {
      opacity: 1;
      transform: scale(1);
      transition: opacity 0.1s, transform 0.1s;
    }
    &:active {
      transform: scale(0.8);
      transition: transform 0.1s;
    }
  `};
`

const MonthTitle = styled.span`
  flex: 1;
  text-align: center;
`

/**
 * Primary UI component for user interaction
 */
export const DatePicker = ({
  className,
  variant = defaultProps.variant,
  selected = defaultProps.selected,
  fillEmptyDays = defaultProps.fillEmptyDays,
  onChange,
  ...rest
}: DatePickerProps) => {
  const [luxonSelectedDay, setLuxonSelectedDay] = useState(DateTime.fromJSDate(selected).startOf('day'))
  const [luxonSelectedMonth, setLuxonSelectedMonth] = useState(DateTime.fromJSDate(selected).startOf('day'))

  const handleClickNext = () => {
    const nextLuxonMonth = luxonSelectedMonth.plus({ months: 1 })
    setLuxonSelectedMonth(nextLuxonMonth)
  }

  const handleClickBack = () => {
    const nextLuxonMonth = luxonSelectedMonth.minus({ months: 1 })
    setLuxonSelectedMonth(nextLuxonMonth)
  }

  const hanldeChange = ({ date }: DayArguments) => {
    setLuxonSelectedDay(DateTime.fromJSDate(date).startOf('day'))
    onChange?.({ date })
  }

  return (
    <Wrapper {...{ className }}>
      <Header>
        <IconButton onClick={handleClickBack}>{'<'}</IconButton>
        <MonthTitle>{luxonSelectedMonth.toFormat('MMMM y')}</MonthTitle>
        <IconButton onClick={handleClickNext}>{'>'}</IconButton>
      </Header>
      <MonthDays
        {...{ variant }}
        daySelected={luxonSelectedDay.toJSDate()}
        monthSelected={luxonSelectedMonth.toJSDate()}
        fillEmptyDays={fillEmptyDays}
        onChange={hanldeChange}
      />
    </Wrapper>
  )
}
