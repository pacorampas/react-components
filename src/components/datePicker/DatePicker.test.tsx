import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import { DatePicker } from './DatePicker'

describe('<DatePicker />', () => {
  test('should change the month title', async () => {
    const { getByRole, getByText } = render(<DatePicker date={new Date('2022-09-01')} />)

    const buttonNext = getByRole('button', { name: />/i })
    const buttonBack = getByRole('button', { name: /</i })
    const monthTitle = getByText(/September/i)

    expect(monthTitle).toContainHTML('September 2022')
    fireEvent.click(buttonNext)

    expect(monthTitle).toContainHTML('October 2022')
    fireEvent.click(buttonBack)

    expect(monthTitle).toContainHTML('September 2022')
  })
})
