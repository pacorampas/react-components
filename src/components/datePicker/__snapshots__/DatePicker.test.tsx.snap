// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Day /> with modifiers should render as out of month 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  padding: 0;
  margin: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 30px;
  box-sizing: border-box;
  border-radius: 6px;
  cursor: pointer;
  opacity: 0.5;
  background: transparent;
}

<button
  class="c0"
  tabindex="-1"
>
  1
</button>
`;

exports[`<Day /> with modifiers should render no selected 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  padding: 0;
  margin: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 30px;
  box-sizing: border-box;
  border-radius: 6px;
  cursor: pointer;
  opacity: 1;
  background: transparent;
}

<button
  class="c0"
  tabindex="-1"
>
  1
</button>
`;

exports[`<Day /> with modifiers should render selected 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  padding: 0;
  margin: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 30px;
  box-sizing: border-box;
  border-radius: 6px;
  cursor: pointer;
  opacity: 1;
  background: pink;
  background: pink;
  color: black;
  border-color: pink;
}

.c0:hover {
  background: #ffd4df;
  border-color: #ffd4df;
  color: black;
}

.c0:active {
  background: #ffdee9;
  border-color: #ffdee9;
  color: black;
}

.c0:focus {
  border-color: #e5a7b2;
}

<button
  class="c0"
  tabindex="-1"
>
  1
</button>
`;

exports[`<Day /> with modifiers should render using info 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  padding: 0;
  margin: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 30px;
  box-sizing: border-box;
  border-radius: 6px;
  cursor: pointer;
  opacity: 1;
  background: transparent;
}

<button
  class="c0"
  tabindex="-1"
>
  1
</button>
`;

exports[`<Day /> with modifiers should render using negative 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  padding: 0;
  margin: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 30px;
  box-sizing: border-box;
  border-radius: 6px;
  cursor: pointer;
  opacity: 1;
  background: transparent;
}

<button
  class="c0"
  tabindex="-1"
>
  1
</button>
`;

exports[`<Day /> with modifiers should render using positive 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  padding: 0;
  margin: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 30px;
  box-sizing: border-box;
  border-radius: 6px;
  cursor: pointer;
  opacity: 1;
  background: transparent;
}

<button
  class="c0"
  tabindex="-1"
>
  1
</button>
`;

exports[`<Day /> with modifiers should render using primary 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  padding: 0;
  margin: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 30px;
  box-sizing: border-box;
  border-radius: 6px;
  cursor: pointer;
  opacity: 1;
  background: transparent;
}

<button
  class="c0"
  tabindex="-1"
>
  1
</button>
`;

exports[`<Day /> with modifiers should render using warning 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  padding: 0;
  margin: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 30px;
  box-sizing: border-box;
  border-radius: 6px;
  cursor: pointer;
  opacity: 1;
  background: transparent;
}

<button
  class="c0"
  tabindex="-1"
>
  1
</button>
`;

exports[`<Day /> with modifiers should render with tabindex 0 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 0;
  padding: 0;
  margin: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 30px;
  box-sizing: border-box;
  border-radius: 6px;
  cursor: pointer;
  opacity: 1;
  background: transparent;
}

<button
  class="c0"
  tabindex="0"
>
  1
</button>
`;
